### PR TITLE
Add esm resolver for bundlers

### PR DESCRIPTION
**Pull request by mkslanc**
*Issue #, if available:*

*Description of changes:*
- Add `setModuleLoader` for ace config to allow `ace.config.setModuleLoader('ace/ext/beautify', () => import('./src-noconflict/ext-beautify.js'));` dynamic imports for all extensions for bundlers (tested in Webpack 5, Parcel, Vite).
- Add ESM resolver generator to be called on `prepack` for `ace-code` package (respecting paths)
- Add  ESM resolver generator for `ace-builds` (respecting paths)

I left `webpack-resolver` generator as it could be used for webpack 4 or less. 

Example builds could be found [here](https://github.com/mkslanc/ace-samples)

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/config_test.js [hunk 1]
```diff
@@ -39,6 +39,10 @@ module.exports = {
         url = config.moduleUrl("ace/ext/textarea");
         assert.equal(url, "a/b/ext-textarea.js");
         config.set("packaged", false);
+        
+        var callback = () => "testCallback";
+        config.setModuleLoader("ace/test-module", callback);
+        assert.equal(config.dynamicModules["ace/test-module"], callback);
```
Comment (nightwing) [inline src/config_test.js:45]: Could you expand this test to also call loadModule and check the result?

File: tool/esm_resolver_generator.js [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    let keyBindingsPaths = ["vim", "emacs", "sublime", "vscode"].map(function (name) {
+        return "ace/keyboard/" + name;
+    });
+    // extensions
+    let extPaths = jsFileList("src/ext").map(function (name) {
+        return "ace/ext/" + name;
+    });
+    // workers
+    let workersPath = workers("src/mode").map(function (name) {
+        return "src/mode/" + name + "_worker";
```
Comment (nightwing) [inline tool/esm_resolver_generator.js:47]: do we need this for ace-code?

File: tool/esm_resolver_generator.js [hunk 3]
```diff
@@ -0,0 +1,65 @@
+var fs = require("fs");
+
+function buildResolver() {
```
Comment (nightwing) [inline tool/esm_resolver_generator.js:3]: can you import these functions from Makefile.dryice.js?

File: src/config.js [hunk 4]
```diff
@@ -85,6 +85,7 @@ exports.setLoader = function(cb) {
     loader = cb;
 };
 
+exports.dynamicModules = {};
```
Comment (nightwing) [inline src/config.js:88]: use `Object.create(null)`, just in case

File: ace.d.ts [hunk 5]
```diff
@@ -371,6 +371,7 @@ export namespace Ace {
     all(): { [key: string]: any };
     moduleUrl(name: string, component?: string): string;
     setModuleUrl(name: string, subst: string): string;
+    setModuleLoader(name: string, onLoad: Function): void;
```
Comment (nightwing) [inline ace.d.ts:374]: looks like setLoader is missing here too

**Review by nightwing — APPROVED** [review]

File: src/config_test.js [hunk 6]
```diff
@@ -39,6 +39,14 @@ module.exports = {
         url = config.moduleUrl("ace/ext/textarea");
         assert.equal(url, "a/b/ext-textarea.js");
         config.set("packaged", false);
+
+        var callback = () => Promise.resolve("success");
```
Comment (nightwing) [inline src/config_test.js:43]: Add `/**global Promise*/` comment to disable the warning.

Comment (mkslanc) [reply src/config_test.js:44]: Done!

File: ace.d.ts (hunk 5 above)
Comment (mkslanc) [reply ace.d.ts:375]: Added

File: tool/esm_resolver_generator.js (hunk 3 above)
Comment (mkslanc) [reply tool/esm_resolver_generator.js:4]: I changed code to import, hope that would work

File: tool/esm_resolver_generator.js (hunk 2 above)
Comment (mkslanc) [reply tool/esm_resolver_generator.js]: You are right, we don't

File: src/config_test.js (hunk 1 above)
Comment (mkslanc) [reply src/config_test.js:46]: Done!

File: src/config_test.js (hunk 6 above)
Comment (mkslanc) [reply src/config_test.js:44]: Seems, didn't helped, I could use some mock construction here to avoid this

**Pull request merged**